sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -0,0 +1,358 @@
+        extra_client_params=extra_client_params)
+
+  def expand(self, pcoll):
+    return pcoll | iobase.Read(self._mongo_source)
+
+
+class _BoundedMongoSource(iobase.BoundedSource):
+  def __init__(self,
+               uri=None,
+               db=None,
+               coll=None,
+               filter=None,
+               projection=None,
+               extra_client_params=None):
+    if extra_client_params is None:
+      extra_client_params = {}
+    if filter is None:
+      filter = {}
+    self.uri = uri
+    self.db = db
+    self.coll = coll
+    self.filter = filter
+    self.projection = projection
+    self.spec = extra_client_params
+    self.doc_count = self._get_document_count()
+    self.avg_doc_size = self._get_avg_document_size()
+    self.client = None
+
+  def estimate_size(self):
+    return self.avg_doc_size * self.doc_count
+
+  def split(self, desired_bundle_size, start_position=None, stop_position=None):
+    # use document cursor index as the start and stop positions
+    if start_position is None:
+      start_position = 0
+    if stop_position is None:
+      stop_position = self.doc_count
+
+    # get an estimate on how many documents should be included in a split batch
+    desired_bundle_count = desired_bundle_size // self.avg_doc_size
+
+    bundle_start = start_position
+    while bundle_start < stop_position:
+      bundle_end = min(stop_position, bundle_start + desired_bundle_count)
+      yield iobase.SourceBundle(weight=bundle_end - bundle_start,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=bundle_end)
+      bundle_start = bundle_end
+
+  def get_range_tracker(self, start_position, stop_position):
+    if start_position is None:
+      start_position = 0
+    if stop_position is None:
+      stop_position = self.doc_count
+    return OffsetRangeTracker(start_position, stop_position)
+
+  def read(self, range_tracker):
+    with MongoClient(self.uri, **self.spec) as client:
+      # docs is a MongoDB Cursor
+      docs = client[self.db][self.coll].find(
+          filter=self.filter, projection=self.projection
+      )[range_tracker.start_position():range_tracker.stop_position()]
+      for index in range(range_tracker.start_position(),
+                         range_tracker.stop_position()):
+        if not range_tracker.try_claim(index):
+          return
+        yield docs[index - range_tracker.start_position()]
+
+  def display_data(self):
+    res = super(_BoundedMongoSource, self).display_data()
+    res['uri'] = self.uri
+    res['database'] = self.db
+    res['collection'] = self.coll
+    res['filter'] = self.filter
+    res['project'] = self.projection
+    res['mongo_client_spec'] = self.spec
+    return res
+
+  def _get_avg_document_size(self):
+    with MongoClient(self.uri, **self.spec) as client:
+      return client[self.db].command('collstats', self.coll)['avgObjSize']
```

Comment:
Will we be always get a valid size here ? 

Otherwise we should validate and pick a suitable default (for example if service return a negative value for some reason).

Reply:
it may be hard to pick a suitable value, and could cause unwanted side effect, I can validate the value and throw exception if the value is unexpected.